MdeModulePkg/Universal/Console/TerminalDxe/Terminal.c

```diff
@@ -385,10 +386,19 @@ StartTerminalStateMachine (
                   );
   ASSERT_EFI_ERROR (Status);
 
+  if (PcdGet32 (PcdTerminalKeyboardTimerInterval) != 0) {
+    TimerInterval = PcdGet32 (PcdTerminalKeyboardTimerInterval);
+  } else if (TerminalDevice->SerialInTimeOut == 0) {
+    TimerInterval = KEYBOARD_TIMER_INTERVAL;
+  } else {
+    // We should check back when filled in half of the FIFO, the number is the unit of 100ns.
+    TimerInterval = TerminalDevice->SerialInTimeOut * FIFO_MAX_NUMBER / 2 * 10;

```

Comment:
Some UARTs  do not have a FIFO or may be configured with the FIFO disabled.  Also, UARTs have have very different FIFO depths.

Reply:
Thanks Mike for the feedback.

I'm going to close this. I think this needs a deeper rework to make it compatible with the various serial devices.